### ISSUE TITLE
Setting to disable the telescope

### ISSUE THREAD

**Issue by AlexAltea** [pwndbg/pwndbg#371]
I use pwndbg to debug an operating system and some device drivers running under QEMU. The problem is that while doing so, some registers may hold MMIO register addresses which the telescope attempts to dereference. This not only causes extra annoying logs, but for several reasons also creates non-determinism despite issuing only read queries.

This is similar to the 2nd point mentioned in issue https://github.com/pwndbg/pwndbg/pull/369, and might (?) be fixed by https://github.com/pwndbg/pwndbg/pull/367. If I understood that pull request correctly, I should be able to enter `set dereference-limit 0` to disable the telescope.

**Comment by disconnect3d** [comment]
@AlexAltea the #367 has been merged.  Does that work for you?

PS: Might be worth to use code from https://github.com/pwndbg/pwndbg/pull/377as #369 made getting stack upper bound wrong (we need tests ASAP!).

**Comment by AlexAltea** [comment]
Although #367 does its job, it does not solve my problem. Despite setting `set dereference-limit 0`, I still get memory reads (unrelated to the code that it's being debugged) while stepping through the code. I'm not sure what is causing that nor how to debug this issue.

I can only confirm that on stock GDB on TUI mode with asm+regs layouts that issue doesn't happen.

**Comment by disconnect3d** [comment]
@AlexAltea Could you prepare some minimal working example that reproduces the error so we could see that (and add a testcase for it in the future)?

**Comment by AlexAltea** [comment]
@disconnect3d I've been debugging the issue myself, since it's hard making a minimal test case, and sending you a 20 GB .ovf file of pre-configured VM would be a little too extreme. :-)

Turns out:
- `set dereference-limit 0` (from #367) does indeed fix some unwanted reads.
- `set emulate off` fixes more unwanted reads triggered by Unicorn.

But there's still few unwanted memory reads that I wasn't able to remove. Here's the function call stack that leads to them:

```
File "/home/user/Desktop/pwndbg/pwndbg/prompt.py", line 37, in prompt_hook
    prompt_hook_on_stop(*a)
File "/home/user/Desktop/pwndbg/pwndbg/memoize.py", line 48, in __call__
    value = self.func(*args, **kwargs)
File "/home/user/Desktop/pwndbg/pwndbg/prompt.py", line 42, in prompt_hook_on_stop
    pwndbg.commands.context.context()
File "/home/user/Desktop/pwndbg/pwndbg/commands/__init__.py", line 106, in __call__
    return self.function(*args, **kwargs)
File "/home/user/Desktop/pwndbg/pwndbg/commands/__init__.py", line 197, in _OnlyWhenRunning
    return function(*a, **kw)
File "/home/user/Desktop/pwndbg/pwndbg/commands/context.py", line 64, in context
    result.extend(func())
File "/home/user/Desktop/pwndbg/pwndbg/commands/context.py", line 76, in context_regs
    return [pwndbg.ui.banner("registers")] + get_regs()
File "/home/user/Desktop/pwndbg/pwndbg/commands/context.py", line 120, in get_regs
    desc = pwndbg.chain.format(value)
File "/home/user/Desktop/pwndbg/pwndbg/chain.py", line 84, in format
    chain = get(value, limit, offset, hard_stop, hard_end)
File "/home/user/Desktop/pwndbg/pwndbg/chain.py", line 49, in get
    address = int(pwndbg.memory.poi(pwndbg.typeinfo.ppvoid, address + offset))
File "/home/user/Desktop/pwndbg/pwndbg/memory.py", line 357, in poi
    for line in traceback.format_stack():
```

For some strange reason I was able to remove all uncessary reads with `set dereference-limit -1`, but that  (obviously) totally destroys the rendering of registers:

```
*RAX  add    byte ptr [rax], al /* 0xffffafe1d0340000 */
 RBX  1
 RCX  add    byte ptr [rax], dl /* 0xffffafe1d0341000 */
 RDX  add    byte ptr [rax], al /* 0xffffafe1d0340000 */
[...]
 R15  0xffff800000000000
 RBP  movabs al, byte ptr [0xa0ffffffff8456b5] /* 0xffffffff8456b5a0 */
 RSP  loopne 0xffffffff8456b5d4 /* 0xffffffff836b62e0 */
*RIP  mov    rdx, rcx /* 0x48d0440f48ca8948 */
```

Is this useful enough? I can try to provide any other details if needed. Thanks a lot for your help.

**Comment by AlexAltea** [comment]
Ok, I think I've found the issue. Happens always after commenting in GitHub. :-)

Notice this code in `pwndbg/chain.py`:

```python
def get(address, limit=LIMIT, offset=0, hard_stop=None, hard_end=0):
[...]
    limit = int(limit)+1

[...]

def format(value, limit=LIMIT, code=True, offset=0, hard_stop=None, hard_end=0):
[...]
    limit = int(limit)+1
[...]
        chain = get(value, limit, offset, hard_stop, hard_end)
```

When `format` calls `get` you increment `limit` twice.

**Comment by disconnect3d** [comment]
We can try to debug and fix the issue through a teamviewer if you want. Can
you join our IRC channel on freenode?

23.12.2017 7:38 PM "Alexandro Sanchez" <notifications@github.com>
napisał(a):

**Comment by disconnect3d** [comment]
Guess this can be closed as #380 has been merged.
